platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameHandler.kt

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.cc.impl.problems
+
+import org.gradle.api.initialization.Settings
+import org.gradle.api.internal.SettingsInternal
+import org.gradle.internal.InternalBuildAdapter
+
+class BuildNameHandler : InternalBuildAdapter() {
```

Comment:
❓  What is the reason we want to extract this?

Reply:
I'm reusing this for problems report generation

Reply:
Then I think it should be turned into an independent service that provides the build name once it is available. 

We should avoid having classes that need manual lifecycle management if their registration is unconditional.

At the same time, there are two approaches to this now: explicit and implicit `ListenerManager` registration. We don't have a guideline on which one is a best practice.

I think you should go with the explicit one for now. See an example here:
https://github.com/gradle/gradle/blob/94faa3d523951f395123f459d4f513bd417a654f/platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinCompilerContextDisposer.kt

Given it becomes a service, I'd
- Change its name to something more appropriate like `BuildNameProvider`
- Add a javadoc
- Hide the `bulidName` field and throw an exception in case the build name is requested before it was initialized